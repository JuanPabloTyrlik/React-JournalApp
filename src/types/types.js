export const TYPES = {
    // Auth

    LOGIN: '[Auth] Login',
    LOGOUT: '[Auth] Logout',

    // UI

    UI_SET_ERROR: '[UI] SetError',
    UI_REMOVE_ERROR: '[UI] RemoveError',
    UI_START_LOADING: '[UI] Start Loading',
    UI_FINISH_LOADING: '[UI] Finish Loading',
};
