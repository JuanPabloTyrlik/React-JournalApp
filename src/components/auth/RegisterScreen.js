import React from 'react';
import validator from 'validator';
import { Link } from 'react-router-dom';
import Alert from '@material-ui/lab/Alert';
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
    const [values, handleChange] = useForm({
        name: 'Juan',
        email: 'juan@gmail.com',
        password: '123456',
        confirmPassword: '123456',
    });

    const { name, email, password, confirmPassword } = values;

    const dispatch = useDispatch();

    const ui = useSelector((state) => state.ui);

    const handleRegister = (e) => {
        e.preventDefault();
        e.stopPropagation();
        if (isValidForm()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }
    };

    const isValidForm = () => {
        if (/^\s*$/.test(name)) {
            dispatch(setError('Name cannot be empty'));
            return false;
        }
        if (!validator.isEmail(email)) {
            dispatch(setError('Email must be valid'));
            return false;
        }
        if (password.trim().length <= 5) {
            dispatch(setError('Password must be at least 6 characters'));
            return false;
        }
        if (password !== confirmPassword) {
            dispatch(setError("Passwords don't match"));
            return false;
        }
        dispatch(removeError());
        return true;
    };

    const handleRedirect = () => {
        dispatch(removeError());
    };

    return (
        <>
            <h1 className="auth__title">Register</h1>
            {ui.msgError && (
                <Alert severity="error" style={{ marginBottom: '1rem' }}>
                    {ui.msgError}
                </Alert>
            )}
            <form onSubmit={handleRegister}>
                <input
                    type="text"
                    name="name"
                    id="name"
                    placeholder="Name"
                    autoComplete="off"
                    className="auth__input"
                    value={name}
                    onChange={handleChange}
                />
                <input
                    type="text"
                    name="email"
                    id="email"
                    placeholder="Email"
                    autoComplete="off"
                    className="auth__input"
                    value={email}
                    onChange={handleChange}
                />
                <input
                    type="password"
                    name="password"
                    id="password"
                    placeholder="Password"
                    className="auth__input"
                    value={password}
                    onChange={handleChange}
                />
                <input
                    type="password"
                    name="confirmPassword"
                    id="confirmPassword"
                    placeholder="Confirm Password"
                    className="auth__input"
                    value={confirmPassword}
                    onChange={handleChange}
                />
                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={ui.loading}
                >
                    Create account
                </button>
                <div className="mt-1">
                    <Link
                        to="/auth/login"
                        className="link"
                        onClick={handleRedirect}
                    >
                        Already have an account?
                    </Link>
                </div>
            </form>
        </>
    );
};
