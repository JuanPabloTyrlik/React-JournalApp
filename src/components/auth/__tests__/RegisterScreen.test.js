import React from 'react';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RegisterScreen } from '../RegisterScreen';
import { TYPES } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const store = mockStore({
    ui: {
        loading: false,
        msgError: null,
    },
    auth: {},
    notes: {
        notes: [],
        active: null,
    },
});

describe('Tests on <RegisterScreen />', () => {
    beforeEach(() => {
        store.clearActions();
        jest.clearAllMocks();
    });
    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
    );
    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('should update error messages', () => {
        const form = wrapper.find('form');
        form.simulate('submit');
        expect(store.getActions()[0]).toEqual({
            type: TYPES.UI_SET_ERROR,
            payload: 'Name cannot be empty',
        });
        wrapper.find('input[name="name"]').simulate('change', {
            target: { name: 'name', value: 'Test User' },
        });
        form.simulate('submit');
        expect(store.getActions()[1]).toEqual({
            type: TYPES.UI_SET_ERROR,
            payload: 'Email must be valid',
        });
        wrapper.find('#email').simulate('change', {
            target: {
                name: 'email',
                value: 'test.user@testing.com',
            },
        });
        form.simulate('submit');
        expect(store.getActions()[2]).toEqual({
            type: TYPES.UI_SET_ERROR,
            payload: 'Password must be at least 6 characters',
        });
        wrapper.find('#password').simulate('change', {
            target: {
                name: 'password',
                value: '123456',
            },
        });
        form.simulate('submit');
        expect(store.getActions()[3]).toEqual({
            type: TYPES.UI_SET_ERROR,
            payload: "Passwords don't match",
        });
    });
    test('should render error messages', () => {
        const store = mockStore({
            ui: {
                loading: false,
                msgError: 'Email must be valid',
            },
        });
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
        expect(wrapper.find('.MuiAlert-message').text()).toBe(
            'Email must be valid'
        );
    });
});
