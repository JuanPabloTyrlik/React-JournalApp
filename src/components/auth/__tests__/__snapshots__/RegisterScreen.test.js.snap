// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <RegisterScreen /> should render correctly 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h1
    className="auth__title"
  >
    Register
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      id="confirmPassword"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Create account
    </button>
    <div
      className="mt-1"
    >
      <a
        className="link"
        href="/auth/login"
        onClick={[Function]}
      >
        Already have an account?
      </a>
    </div>
  </form>
</div>
`;
